crates/ty_python_semantic/resources/mdtest/public_types.md

```diff
@@ -0,0 +1,421 @@
+# Public types
+
+## Basic
+
+The "public type" of a symbol refers to the type that is inferred for a symbol from an enclosing
+scope. Since it is not generally possible to analyze the full control flow of a program, we
+currently make the simplifying assumption that the inner scope (such as the `inner` function below)
+could be executed at any position in the enclosing scope. The public type should therefore be the
+union of all possible types that the symbol could have.
+
+In the following example, depending on when `inner()` is called, the type of `x` could either be `A`
+or `B`:
+
+```py
+class A: ...
+class B: ...
+class C: ...
+
+def outer() -> None:
+    x = A()
+
+    def inner() -> None:
+        reveal_type(x)  # revealed: Unknown | A | B
+    # This call would observe `x` as `A`.
+    inner()
+
+    x = B()
+
+    # This call would observe `x` as `B`.
+    inner()
+```
+
+Similarly, if control flow in the outer scope can split, the public type of `x` should reflect that:
+
+```py
+def outer(flag: bool) -> None:
+    x = A()
+
+    def inner() -> None:
+        reveal_type(x)  # revealed: Unknown | A | B | C
+    inner()
+
+    if flag:
+        x = B()
+
+        inner()
+    else:
+        x = C()
+
+        inner()
+
+    inner()
+```
+
+If a binding is not reachable, it is not considered in the public type:
+
+```py
+def outer() -> None:
+    x = A()
+
+    def inner() -> None:
+        reveal_type(x)  # revealed: Unknown | A | C
+    inner()
+
+    if False:
+        x = B()  # this binding of `x` is unreachable
+        inner()
+
+    x = C()
+    inner()
+
+def outer(flag: bool) -> None:
+    x = A()
+
+    def inner() -> None:
+        reveal_type(x)  # revealed: Unknown | A | C
+    inner()
+
+    if flag:
+        return
+
+        x = B()  # this binding of `x` is unreachable
+
+    x = C()
+    inner()
+```
+
+If a symbol is only conditionally bound, we do not raise any errors:
```

Comment:
To more clearly distinguish this case from the case that we _do_ want to complain about (which you detail in your following example):

```suggestion
If a symbol is only conditionally bound and the inner scope is bound in the same branch,
we do not raise any errors:
```

Reply:
sure